include/picongpu/plugins/binning/Binner.hpp

```diff
@@ -0,0 +1,319 @@
+/* Copyright 2023 Tapish Narwal
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/plugins/binning/BinningFunctors.hpp"
+#include "picongpu/plugins/binning/DomainInfo.hpp"
+#include "picongpu/plugins/binning/WriteHist.hpp"
+#include "picongpu/plugins/binning/utility.hpp"
+
+
+namespace picongpu
+{
+    namespace plugins::binning
+    {
+        /**
+         * Functor to run on divide for time averaging
+         */
+        // TODO make this generic? apply a functor on all elements of a databox? take in functor?
+        template<uint32_t blockSize, uint32_t nAxes>
+        struct ProductKernel
+        {
+            using result_type = void;
+
+            HINLINE
+            ProductKernel()
+            {
+            }
+
+            template<typename T_Worker, typename T_DataSpace, typename T_DepositedQuantity, typename T_DataBox>
+            HDINLINE void operator()(
+                const T_Worker& worker,
+                const T_DataSpace& extentsDataspace,
+                const T_DepositedQuantity factor,
+                T_DataBox dataBox) const
+            {
+                auto blockIdx = cupla::blockIdx(worker.getAcc()).x * blockSize;
+                auto forEachElemInDataboxChunk = lockstep::makeForEach<blockSize>(worker);
+                forEachElemInDataboxChunk(
+                    [&](int32_t const linearIdx)
+                    {
+                        int32_t const linearTid = blockIdx + linearIdx;
+                        if(linearTid < extentsDataspace.productOfComponents())
+                            dataBox(DataSpace<1u>{static_cast<int>(linearTid)}) *= factor;
+                    });
+            }
+        };
+
+
+        /**
+         * Functor to do volume normaliztion
+         * Factor with picongpu units
+         * User needs to deal with the units seperately
+         */
+        // TODO make this more generic? databox operations with another databox?
+        // maybe store the normalization in a buffer rather than computing it at every output timestep (memory vs ops)
+        // this style vs having the members as params of operator()
+        template<uint32_t blockSize, uint32_t nAxes>
+        struct BinNormalizationKernel
+        {
+            using result_type = void;
```

Comment:
should be named `ResultType` or `ValueType` to follow our other naming schemas.

Reply:
switched to ResultType